File: src/main/java/no/publishers/controller/PublishersApiImpl.java
package no.publishers.controller;

import io.swagger.annotations.ApiParam;
import no.publishers.generated.model.Publisher;

import no.publishers.service.PublisherService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;

import java.util.List;

import static org.springframework.web.bind.annotation.RequestMethod.GET;

@Controller
public class PublishersApiImpl implements no.publishers.generated.api.PublishersApi{
    private static Logger LOGGER = LoggerFactory.getLogger(PublishersApiImpl.class);

    @Autowired
    private PublisherService publisherService;

    @RequestMapping(value="/ping", method=GET, produces={"text/plain"})
    public ResponseEntity<String> getPing() {
        return ResponseEntity.ok("pong");
    }

    @RequestMapping(value="/ready", method=GET)
    public ResponseEntity getReady() {
        return ResponseEntity.ok().build();
    }

    @Override
    public ResponseEntity<Void> createPublisher(HttpServletRequest httpServletRequest, @ApiParam(required=true) @Valid @RequestBody Publisher publisher) {
        HttpHeaders headers = new HttpHeaders();

        try {
            headers.setLocation(
                ServletUriComponentsBuilder
                    .fromCurrentServletMapping()
                    .path("/publishers/{id}")
                    .build()
                    .expand(publisherService.createPublisher(publisher).getId())
                    .toUri());
        } catch (Exception e) {
            LOGGER.error("createPublisher failed:", e);
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }

        return new ResponseEntity<>(headers, HttpStatus.CREATED);
    }
    
    @Override
    public ResponseEntity<Publisher> getPublisherById(HttpServletRequest httpServletRequest, @ApiParam(value = "id",required=true) @PathVariable("id") String id) {
        ResponseEntity<Publisher> response;

        try {
            response = publisherService.getById(id)
                .map(publisher -> new ResponseEntity<>(publisher, HttpStatus.OK))
                .orElse( new ResponseEntity<>(HttpStatus.NOT_FOUND) );
        } catch (Exception e) {
            LOGGER.error("getPublisherById failed:", e);
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }

        return response;
    }

    @Override
    public ResponseEntity<List<Publisher>> getPublishers(HttpServletRequest httpServletRequest, @ApiParam(value = "A query string to match a publisher name") @Valid @RequestParam(value = "name", required = false) String name,@ApiParam(value = "If you want to filter by organizationId") @Valid @RequestParam(value = "organizationId", required = false) String organizationId) {
        List<Publisher> publishers;

        try {
            publishers = publisherService.getPublishers(name, organizationId);
        } catch (Exception e) {
            LOGGER.error("getPublishers failed:", e);
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }

        return new ResponseEntity<>(publishers, HttpStatus.OK);
    }
}

File: src/main/kotlin/no/publishers/controller/PublishersApiImpl.kt
package no.publishers.controller

import javax.servlet.http.HttpServletRequest
import no.publishers.generated.model.Publisher
import no.publishers.jena.MissingAcceptHeaderException
import no.publishers.jena.jenaResponse
import no.publishers.service.PublisherService
import org.slf4j.LoggerFactory
import org.springframework.dao.DuplicateKeyException
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod.GET
import org.springframework.web.servlet.support.ServletUriComponentsBuilder
import javax.validation.ConstraintViolationException

private val LOGGER = LoggerFactory.getLogger(PublishersApiImpl::class.java)

@Controller
open class PublishersApiImpl (
    private val publisherService: PublisherService
): no.publishers.generated.api.PublishersApi {

    @RequestMapping(value = ["/ping"], method = [GET], produces = ["text/plain"])
    fun ping(): ResponseEntity<String> =
        ResponseEntity.ok("pong")

    @RequestMapping(value = ["/ready"], method = [GET])
    fun ready(): ResponseEntity<Void> =
        ResponseEntity.ok().build()

    override fun createPublisher(httpServletRequest: HttpServletRequest, publisher: Publisher): ResponseEntity<Void> =
        try {
            HttpHeaders()
                .apply {
                    location = ServletUriComponentsBuilder
                        .fromCurrentServletMapping()
                        .path("/publishers/{id}")
                        .build()
                        .expand(publisherService.createPublisher(publisher).id)
                        .toUri() }
                .let { ResponseEntity(it, HttpStatus.CREATED) }
        } catch (exception: Exception) {
            LOGGER.error("createPublisher failed:", exception)
            when(exception) {
                is ConstraintViolationException -> ResponseEntity(HttpStatus.BAD_REQUEST)
                is DuplicateKeyException -> ResponseEntity(HttpStatus.CONFLICT)
                else -> ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR)
            }
        }

    override fun updatePublisher(httpServletRequest: HttpServletRequest, id: String, publisher: Publisher): ResponseEntity<Publisher> =
        try {
            publisherService
                .updatePublisher(id, publisher)
                ?.let { updated -> ResponseEntity(updated, HttpStatus.OK) }
                ?: ResponseEntity(HttpStatus.NOT_FOUND)
        } catch (exception: Exception) {
            LOGGER.error("updatePublisher failed:", exception)
            when(exception) {
                is ConstraintViolationException -> ResponseEntity(HttpStatus.BAD_REQUEST)
                is DuplicateKeyException -> ResponseEntity(HttpStatus.CONFLICT)
                else -> ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR)
            }
        }

    override fun getPublisherById(httpServletRequest: HttpServletRequest, id: String): ResponseEntity<String> =
        try {
            publisherService
                .getById(id)
                ?.let { publisher -> publisher.jenaResponse(httpServletRequest.getHeader("Accept")) }
                ?.let { response -> ResponseEntity(response, HttpStatus.OK) }
                ?: ResponseEntity(HttpStatus.NOT_FOUND)
        } catch (exception: Exception) {
            LOGGER.error("getPublisherById failed:", exception)
            when(exception) {
                is MissingAcceptHeaderException -> ResponseEntity(HttpStatus.NOT_ACCEPTABLE)
                else -> ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR)
            }
        }

    override fun getPublishers(httpServletRequest: HttpServletRequest, name: String?, organizationId: String?): ResponseEntity<String> =
        try {
            publisherService
                .getPublishers(name, organizationId)
                .let { publishers -> publishers.jenaResponse(httpServletRequest.getHeader("Accept")) }
                .let { response -> ResponseEntity(response, HttpStatus.OK) }
        } catch (exception: Exception) {
            LOGGER.error("getPublishers failed:", exception)
            when(exception) {
                is MissingAcceptHeaderException -> ResponseEntity(HttpStatus.NOT_ACCEPTABLE)
                else -> ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR)
            }
        }
}
